### PR TITLE
TST: Review ``_assert_within_tol`` positional arguments

### LINKED ISSUE #15058: TST: review usage of `_assert_within_tol` with positional arguments in `test_distance.py`
While working on #15055, I found that tests in `test_distance.py` (seemingly unintentionally) use absolute rather than relative tolerance, because `_assert_within_tol` is often called with positional parameters that end up being filled into the signature:
```
def _assert_within_tol(a, b, atol=0, rtol=0, verbose_=False):
```

There are some that are definitely misleading (or outright false), because a call-site like
```
_assert_within_tol(Y1, Y2, eps, verbose > 2)
```
will end up calling `rtol=False/True`, which then presumably gets cast to 0/1.

I think these tests need some review if they're working as intended, and @peterbell10 agreed in #15055 that this should be tracked.

### PR DESCRIPTION
**Pull request by Kai-Striega**
#### Reference issue

closes gh-15058
#### What does this implement/fix?

``_assert_within_tol`` is often called with positional arguments that end up being filled into the signature, e.g. calls to ``_assert_within_tol(Y1, Y2, eps, verbose > 2)``  evaluate to calls where `atol=eps` and `rtol=verbose > 2`.

 This PR converts calls to using keywords where needed. Further it modifies some tolerances to be stricter.
#### Additional information

The `eps` values appear to be tailored towards absolute tolerance not relative tolerance. I've made a judgement call as to which is more appropriate for any given test, but I'm happy to change any if the reviewers think they other option is better suited.

There are several instances where ``_assert_within_tol`` is called without ``verbose``. This is equivalent to ``assert_allclose``. I think in these cases it will be clearer to call ``assert_allclose`` directly. I'd be happy to change this but haven't made any changes yet.

### CODE REVIEW

File: scipy/spatial/tests/test_distance.py [hunk 1]
```diff
@@ -575,11 +575,11 @@ def test_cdist_dtype_equivalence(self):
                 else:
                     for new_type in test[1]:
                         y2 = cdist(new_type(X1), new_type(X2), metric=metric)
-                        _assert_within_tol(y1, y2, eps, verbose > 2)
+                        _assert_within_tol(y1, y2, rtol=eps, verbose_=verbose > 2)
```
Comment (h-vetinari) [inline scipy/spatial/tests/test_distance.py:578]: Did you try tightening the `eps` in this test? It's still at `1e-07`.

File: scipy/spatial/tests/test_distance.py [hunk 2]
```diff
@@ -747,7 +747,7 @@ def test_pdist_euclidean_iris_float32(self):
         X = np.float32(eo['iris'])
         Y_right = eo['pdist-euclidean-iris']
         Y_test1 = wpdist_no_const(X, 'euclidean')
-        _assert_within_tol(Y_test1, Y_right, eps, verbose > 2)
+        _assert_within_tol(Y_test1, Y_right, atol=eps, verbose_=verbose > 2)
```
Comment (h-vetinari) [inline scipy/spatial/tests/test_distance.py:750]: Same here.

File: scipy/spatial/tests/test_distance.py [hunk 3]
```diff
@@ -846,7 +846,7 @@ def test_pdist_cosine_iris_float32(self):
         X = np.float32(eo['iris'])
         Y_right = eo['pdist-cosine-iris']
         Y_test1 = wpdist(X, 'cosine')
-        _assert_within_tol(Y_test1, Y_right, eps, verbose > 2)
+        _assert_within_tol(Y_test1, Y_right, atol=eps, verbose_=verbose > 2)
```
Comment (h-vetinari) [inline scipy/spatial/tests/test_distance.py:849]: ... and here...

**Review by h-vetinari — COMMENTED** [review]
Thanks a lot for tackling this @Kai-Striega! In general the approach looks great!

I just have one question (that I stopped repeating in the review after the 3rd time) - IMO all tolerances that were previously used in the absolute should be possible to reduce to `1e-15` or maybe `1e-14` in relative terms. Where that isn't feasible, it might need further investigation, or at least a comment why the relative tolerance needs to be looser than `1e-14`.

**Comment by h-vetinari** [comment]
Regarding your point:
> There are several instances where `_assert_within_tol` is called without `verbose`. This is equivalent to `assert_allclose`. I think in these cases it will be clearer to call `assert_allclose` directly. I'd be happy to change this but haven't made any changes yet.

I think actually there is actually no (more) raison d'être for `_assert_within_tol`. The function `assert_allclose` already has a `verbose` argument that [defaults](https://github.com/numpy/numpy/blob/maintenance/1.21.x/numpy/testing/_private/utils.py#L1476-L1477) to true, which [will](https://github.com/numpy/numpy/blob/maintenance/1.21.x/numpy/testing/_private/utils.py#L1530-L1531) [print](https://github.com/numpy/numpy/blob/maintenance/1.21.x/numpy/testing/_private/utils.py#L818-L838) the maximum difference between the arrays both in absolute as well as relative terms.

In other words the addition of
```
    if verbose_:
        print(np.abs(a - b).max())
```
being done by `_assert_within_tol` is at best useless. IMO that function should be removed wholesale.

**Comment by Kai-Striega** [comment]
> IMO all tolerances that were previously used in the absolute should be possible to reduce to 1e-15 or maybe 1e-14 in relative terms.

There were a couple of instances where `eps` seem to, I think, have been designed for use with `atol`, I left those as is. After that I think forgot to check for further tolerances. I'll get them done today.

> IMO that function should be removed wholesale.

I agree. I'll get onto it today.

**Comment by h-vetinari** [comment]
> There were a couple of instances where `eps` seem to, I think, have been designed for use with `atol`, I left those as is.

Ah, I looked too quickly and overlooked the difference between `atol=` vs. `rtol=`, sorry. In any case, I would suggest that we should first try making all tolerances relative. There are very few cases that justify using (only) absolute tolerance from my POV, and I don't think this test module has any (though I might be wrong).

**Comment by Kai-Striega** [comment]
No worries. I'll convert all of them to `rtol`. As I've said before, I haven't worked with these functions that much, so I'll follow your advice here

**Comment by Kai-Striega** [comment]
@h-vetinari could you please take a look at the following test, I believe these are legitimate failures. As the `rtol` is approx 1for each of these. You'll need to run them with the `-m full` option as they are marked slow.

```
FAILED scipy/spatial/tests/test_distance.py::TestPdist::test_pdist_cosine_iris - AssertionError: 
FAILED scipy/spatial/tests/test_distance.py::TestPdist::test_pdist_cosine_iris_float32 - Assertion...
FAILED scipy/spatial/tests/test_distance.py::TestPdist::test_pdist_cosine_iris_nonC - AssertionErr.
```

**Comment by h-vetinari** [comment]
@Kai-Striega, thanks. I don't see them in CI here, I guess you ran this locally?

If the arrays being compared contain values that are very very close to zero, the relative tolerance can becomes very bad indeed, but I looked at https://github.com/scipy/scipy/blob/master/scipy/spatial/tests/data/pdist-cosine-ml-iris.txt, and the minimum there is ~5e-6. Given the scale of the numbers, the errors _should_ however have been showing up before already? I guess we might not be running `-m full` anywhere in CI?

So it looks like something is definitely off here. There might be an issue with the cosine metric or with those tests.

**Comment by Kai-Striega** [comment]
> I don't see them in CI here, I guess you ran this locally?

You are right, I am running this locally *after* I've changed the test from `atol` to `rtol` for which it is now failing. Here is the full failure output on from my system:

```python
============================================= FAILURES ==============================================
_________________________________ TestPdist.test_pdist_cosine_iris __________________________________
scipy/spatial/tests/test_distance.py:841: in test_pdist_cosine_iris
    _assert_within_tol(Y_test1, Y_right, rtol=eps)
        X          = array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
  ...],
       [6.3, 2.5, 5. , 1.9],
       [6.5, 3. , 5.2, 2. ],
       [6.2, 3.4, 5.4, 2.3],
       [5.9, 3. , 5.1, 1.8]])
        Y_right    = array([1.4208365e-03, 1.2652718e-05, 8.9939315e-04, ..., 2.1952710e-03,
       8.7416055e-04, 1.1307517e-03])
        Y_test1    = array([1.42083650e-03, 1.26527175e-05, 8.99393151e-04, ...,
       2.19527098e-03, 8.74160547e-04, 1.13075175e-03])
        eps        = 1e-15
        self       = <scipy.spatial.tests.test_distance.TestPdist object at 0x7fab452affd0>
scipy/spatial/tests/test_distance.py:228: in _assert_within_tol
    assert_allclose(a, b, rtol=rtol, atol=atol)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-15, atol=0
E   
E   Mismatched elements: 11172 / 11175 (100%)
E   Max absolute difference: 4.99524373e-09
E   Max relative difference: 1.
E    x: array([1.420836e-03, 1.265272e-05, 8.993932e-04, ..., 2.195271e-03,
E          8.741605e-04, 1.130752e-03])
E    y: array([1.420836e-03, 1.265272e-05, 8.993932e-04, ..., 2.195271e-03,
E          8.741605e-04, 1.130752e-03])
        a          = array([1.42083650e-03, 1.26527175e-05, 8.99393151e-04, ...,
       2.19527098e-03, 8.74160547e-04, 1.13075175e-03])
        atol       = 0
        b          = array([1.4208365e-03, 1.2652718e-05, 8.9939315e-04, ..., 2.1952710e-03,
       8.7416055e-04, 1.1307517e-03])
        rtol       = 1e-15
        verbose_   = False
_____________________________ TestPdist.test_pdist_cosine_iris_float32 ______________________________
scipy/spatial/tests/test_distance.py:849: in test_pdist_cosine_iris_float32
    _assert_within_tol(Y_test1, Y_right, rtol=eps, verbose_=verbose > 2)
        X          = array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
  ... 2.5, 5. , 1.9],
       [6.5, 3. , 5.2, 2. ],
       [6.2, 3.4, 5.4, 2.3],
       [5.9, 3. , 5.1, 1.8]], dtype=float32)
        Y_right    = array([1.4208365e-03, 1.2652718e-05, 8.9939315e-04, ..., 2.1952710e-03,
       8.7416055e-04, 1.1307517e-03])
        Y_test1    = array([1.42083725e-03, 1.26526543e-05, 8.99393546e-04, ...,
       2.19527259e-03, 8.74160551e-04, 1.13075227e-03])
        eps        = 1e-15
        self       = <scipy.spatial.tests.test_distance.TestPdist object at 0x7fab45324e50>
scipy/spatial/tests/test_distance.py:228: in _assert_within_tol
    assert_allclose(a, b, rtol=rtol, atol=atol)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-15, atol=0
E   
E   Mismatched elements: 11175 / 11175 (100%)
E   Max absolute difference: 2.01499862e-08
E   Max relative difference: 1.00000004
E    x: array([1.420837e-03, 1.265265e-05, 8.993935e-04, ..., 2.195273e-03,
E          8.741606e-04, 1.130752e-03])
E    y: array([1.420836e-03, 1.265272e-05, 8.993932e-04, ..., 2.195271e-03,
E          8.741605e-04, 1.130752e-03])
        a          = array([1.42083725e-03, 1.26526543e-05, 8.99393546e-04, ...,
       2.19527259e-03, 8.74160551e-04, 1.13075227e-03])
        atol       = 0
        b          = array([1.4208365e-03, 1.2652718e-05, 8.9939315e-04, ..., 2.1952710e-03,
       8.7416055e-04, 1.1307517e-03])
        rtol       = 1e-15
        verbose_   = False
_______________________________ TestPdist.test_pdist_cosine_iris_nonC _______________________________
scipy/spatial/tests/test_distance.py:857: in test_pdist_cosine_iris_nonC
    _assert_within_tol(Y_test2, Y_right, rtol=eps)
        X          = array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
  ...],
       [6.3, 2.5, 5. , 1.9],
       [6.5, 3. , 5.2, 2. ],
       [6.2, 3.4, 5.4, 2.3],
       [5.9, 3. , 5.1, 1.8]])
        Y_right    = array([1.4208365e-03, 1.2652718e-05, 8.9939315e-04, ..., 2.1952710e-03,
       8.7416055e-04, 1.1307517e-03])
        Y_test2    = array([1.42083650e-03, 1.26527175e-05, 8.99393151e-04, ...,
       2.19527098e-03, 8.74160547e-04, 1.13075175e-03])
        eps        = 1e-15
        self       = <scipy.spatial.tests.test_distance.TestPdist object at 0x7fab452b7b50>
scipy/spatial/tests/test_distance.py:228: in _assert_within_tol
    assert_allclose(a, b, rtol=rtol, atol=atol)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-15, atol=0
E   
E   Mismatched elements: 11172 / 11175 (100%)
E   Max absolute difference: 4.99524373e-09
E   Max relative difference: 1.
E    x: array([1.420836e-03, 1.265272e-05, 8.993932e-04, ..., 2.195271e-03,
E          8.741605e-04, 1.130752e-03])
E    y: array([1.420836e-03, 1.265272e-05, 8.993932e-04, ..., 2.195271e-03,
E          8.741605e-04, 1.130752e-03])
        a          = array([1.42083650e-03, 1.26527175e-05, 8.99393151e-04, ...,
       2.19527098e-03, 8.74160547e-04, 1.13075175e-03])
        atol       = 0
        b          = array([1.4208365e-03, 1.2652718e-05, 8.9939315e-04, ..., 2.1952710e-03,
       8.7416055e-04, 1.1307517e-03])
        rtol       = 1e-15
        verbose_   = False
====================================== short test summary info ======================================
FAILED scipy/spatial/tests/test_distance.py::TestPdist::test_pdist_cosine_iris - AssertionError: 
FAILED scipy/spatial/tests/test_distance.py::TestPdist::test_pdist_cosine_iris_float32 - Assertion...
FAILED scipy/spatial/tests/test_distance.py::TestPdist::test_pdist_cosine_iris_nonC - AssertionErr...
```

**Comment by h-vetinari** [comment]
What I don't understand is that something like
```
E   Max absolute difference: 4.99524373e-09
E   Max relative difference: 1.
```
must mean that there's a zero somewhere (due to rel-diff=1), and that the value it _should_ be has magnitue less than 5e-9 (due to abs-diff).

But I checked one of the files (`pdist-cosine-ml-iris.txt`) and the minimal value (which we're comparing against) is 5e-6 - 1000x larger than 5e-9. In other words, I don't understand how this can happen - not all of these constraints can simultaneously hold. Clearly I'm overlooking something, but I don't know what...

**Comment by ilayn** [comment]
`rtol` bounds are quite difficult to debug and it's the reason why we tend to choose `atol` instead. If you look at the compared values they are in the order of 1e-3 and relative tolerance of 1e-15 is basically beyond float eps which is around 2.2e-16. So trying to guess `rtol`s based on the values is difficult to maintain and hence the crude `atol` covering is much simpler.

**Comment by h-vetinari** [comment]
> If you look at the compared values they are in the order of 1e-3 and relative tolerance of 1e-15 is basically beyond float eps which is around 2.2e-16

That's not how `eps` works, which itself is about _relative_ accuracy. Values much smaller than `1e-3` can still be accurately resolved to an accuracy of `eps=1.1e-16` without issue.

Basically, unless the problems are extremely ill-conditioned, we should never be losing more than about 1-2 orders of accuracy relative to `eps`. Of course, arbitrarily badly conditioned _problems_ exist (e.g. huge spread of magnitude of eigenvalues in a matrix).

Since the cosine distance involves summing signed elements, the cancellation _can_ be arbitrarily bad. I haven't looked at the implementation, but if it just naively sums an array, that already can lead to very suboptimal behaviour based on the placement of the values. It would be interesting to see if we could regain some accuracy here by adapting the implementation to sort (yes, I know, that's not necessarily cheap...), and then do the following:
- sum the positive parts in ascending order of magnitude
- sum the negative parts in ascending order of magnitude
- finally, add the two numbers

This would minimize the amount of cancellation as much as possible - of course, this doesn't have to happen in this PR, but just food for thought.

**Comment by ilayn** [comment]
I don't understand. You are using `assert_allclose` which does `abs(x-y) <= atol + rtol*abs(y)` with `rtol=1e-15`. What relative eps do you have in mind? It's a very well defined absolute quantity which is 2**-53 for floats

**Comment by h-vetinari** [comment]
> It's a very well defined absolute quantity which is 2**-53 for floats.

I hate to sound so definite and don't want to be rude, but that is wrong. It is a _relative_  quantity. See e.g. [here](https://en.wikipedia.org/wiki/Floating_point#Machine_precision_and_backward_error_analysis)

**Comment by ilayn** [comment]
I also don't want to start a discussion but respectfully, that relative means relative to the number of bits of the floating point system not to the number you are trying to test against in a rtol * x. So if you have a number x and you try to represent in floating point that is the maximum error you can make in bit representation. It has nothing to do what we are discussing here.

**Comment by h-vetinari** [comment]
> that relative means relative to the number of bits of the floating point system not to the number you are trying to test against in a rtol * x

To the extent that I understand this sentence, I disagree. The "52" in the 2**-52 comes from the significand in the [double format](https://en.wikipedia.org/wiki/Double-precision_floating-point_format#IEEE_754_double-precision_binary_floating-point_format:_binary64), which has 53 bits, all of which are unaffected by the magnitude of the exponent (unless we're at the very bottom of the scale using subnormals).

Since the exponent takes care of the magnitude, this is the accuracy that is _relative to `x`_, where `x` is any number roughly between 2**-1022 and 2**1023.

**Comment by ilayn** [comment]
```python
x = np.random.rand(100)*1e-15
y = x*(1+ np.finfo(float).eps)
np.testing.assert_allclose(x, y, atol=0., rtol=1e-15) # passes

np.testing.assert_allclose(x, y, atol=0., rtol=1e-16)
Traceback (most recent call last):

  File "<ipython-input-25-21a0476d9d64>", line 1, in <module>
    np.testing.assert_allclose(x, y, atol=0., rtol=1e-16)

  File "c:\users\ilhan polat\appdata\local\programs\python\python39\lib\site-packages\numpy-1.22.0.dev0+1754.ge7f773cf4-py3.9-win-amd64.egg\numpy\testing\_private\utils.py", line 1530, in assert_allclose
    assert_array_compare(compare, actual, desired, err_msg=str(err_msg),

  File "c:\users\ilhan polat\appdata\local\programs\python\python39\lib\site-packages\numpy-1.22.0.dev0+1754.ge7f773cf4-py3.9-win-amd64.egg\numpy\testing\_private\utils.py", line 844, in assert_array_compare
    raise AssertionError(msg)

AssertionError: 
Not equal to tolerance rtol=1e-16, atol=0

Mismatched elements: 100 / 100 (100%)
Max absolute difference: 1.97215226e-31
Max relative difference: 2.94964051e-16
 x: array([8.356115e-16, 9.025557e-16, 3.959629e-16, 8.292111e-16,
       7.397840e-16, 7.307387e-16, 5.089460e-16, 3.749618e-16,
       9.118591e-16, 1.213508e-16, 1.724919e-16, 9.150482e-16,...
 y: array([8.356115e-16, 9.025557e-16, 3.959629e-16, 8.292111e-16,
       7.397840e-16, 7.307387e-16, 5.089460e-16, 3.749618e-16,
       9.118591e-16, 1.213508e-16, 1.724919e-16, 9.150482e-16,...

```

Look at the absolute difference.

**Comment by h-vetinari** [comment]
> Look at the absolute difference.

We might be in violent agreement, actually?

The absolute difference is reduced by the magnitude `1e-15` that you're multiplying everything with (roughly; 15+16=31, in terms of negative exponents), and the relative difference remains in the ballpark of `1e-16`, which is exactly what I'm saying.

In other words, there's no reason we shouldn't be able to get within an `rtol=1e-14` of a number like `1e-3`, if we can do so even for (absolute) values of `1e-15`, as you just demonstrated.

**Comment by ilayn** [comment]
Yes so `eps` is working as a well-defined scalar and a fixed number as we agreed on. In fact that's why we have a number coming out of `np.finfo('d').eps` because it is fixed. So a single number can be represented up to eps. That is what we agree indeed. 

And since we know that the errors typically grows linearly with dimension in array manipulations we should at least expect to make `eps x n = O(1e-14)` just by doing artihmetics on arrays say summing or norm computation. Here we have a full 2D array so there is no reason why such high performance should come out from these tests and that's why comparison at such level makes no sense here.

**Comment by h-vetinari** [comment]
> Here we have a full 2D array so there is no reason why such high performance should come out from these tests and that's why comparison at such level makes no sense here.

Sure, it doesn't have to be 1e-14, but currently the relative error is 1e0, i.e. off by 16 orders of magnitude. That definitely sounds off to me.

**Comment by h-vetinari** [comment]
OK, to get this in before 1.8 branches next Monday, how about we only increase the accuracy where it works, and leave the rest for later?

I'd prefer to have this in without the fine-tuning, rather than not at all. Tentatively adding the milestone.

**Comment by Kai-Striega** [comment]
> I'd prefer to have this in without the fine-tuning, rather than not at all. 

I agree. I've "fixed" the tolerances and removed the ``_assert_within_tol`` helper function, so there is at least some improvement. I've marked this as ready to merge. If all the tests pass, @h-vetinari would you be willing to merge it?

**Review by h-vetinari — APPROVED** [review]
> @h-vetinari would you be willing to merge it?

I cannot merge, but I can approve at least. :)

Thanks!

**Comment by Kai-Striega** [comment]
Ok I'll merge it once the tests pass. Thank you for taking the time to review
